filebeat/input/filestream/internal/input-logfile/store.go

```diff
@@ -294,6 +303,29 @@ func (s *store) writeState(r *resource) {
 	}
 }
 
+// resetCursor sets the cursor to the value in cur in the persistent store and
+// drops all pending cursor operations.
+func (s *store) resetCursor(key string, cur interface{}) error {
+	r := s.ephemeralStore.Find(key, false)
+	if r == nil {
+		return fmt.Errorf("resource '%s' not found", key)
+	}
+	defer r.Release()
+
+	r.stateMutex.Lock()
+	defer r.stateMutex.Unlock()
+
+	r.version++
+	r.UpdatesReleaseN(r.activeCursorOperations)
+	r.activeCursorOperations = 0
+	r.pendingCursor = nil
+	r.cursor = cur
```

Comment:
On update/reset the API allows users to only update a subset of fields. This is why we use `typeconv.Convert(&resource.cursor, X)` when updating/overwriting the cursor.